cdap/resource_streaming_program_run.go

```diff
@@ -0,0 +1,332 @@
+					return "spark", nil
+				},
+			},
+			"runtime_arguments": {
+				Type:        schema.TypeMap,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The runtime arguments used to start the program",
+				Elem:        &schema.Schema{Type: schema.TypeString},
+			},
+		},
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(20 * time.Minute),
+			Delete: schema.DefaultTimeout(time.Hour), // This gives the pipeline time to drain processing if in-flight records
+		},
+	}
+}
+
+func resourceStreamingProgramRunCreate(d *schema.ResourceData, m interface{}) error {
+	config := m.(*Config)
+
+	addr := getProgramAddr(config, d)
+	startAddr := urlJoin(addr, "start")
+	runsAddr := urlJoin(addr, "runs")
+
+	argsObj := make(map[string]string)
+	// cast map[string]interface{} to map[string]string
+	for k, val := range d.Get("runtime_arguments").(map[string]interface{}) {
+		argsObj[k] = val.(string)
+	}
+
+	runID, err := uuid.NewRandom()
+	if err != nil {
+		return fmt.Errorf("error generating uuid for faux run id: %v", err)
+	}
+	// This runtime arg will be unused by the pipeline but will allow the provider to associate a run with this resource.
+	argsObj[fauxRunID] = runID.String()
+
+	b, err := json.Marshal(argsObj)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, startAddr, bytes.NewReader(b))
+	if err != nil {
+		return err
+	}
+
+	if _, err := httpCall(config.httpClient, req); err != nil {
+		return err
+	}
+
+	// Poll until actually reaches RUNNING state.
+	return resource.Retry(d.Timeout(schema.TimeoutCreate), func() *resource.RetryError {
+		time.Sleep(10 * time.Second) // avoid spamming retries and initial failure to find run.
+		isRunning, err := isFauxRunIDRunningYet(config, runsAddr, runID.String())
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+		if isRunning {
+			d.SetId(runID.String())
+			return nil
+		}
+		return resource.RetryableError(fmt.Errorf("still waiting for program run with faux run id: %v which is in an initializing state", runID.String()))
+	})
+}
+
+//TODO ?
+func resourceStreamingProgramRunRead(d *schema.ResourceData, m interface{}) error {
+	return nil
+}
+
+type runtimeArgs struct {
+	FauxRunID string `json:"__FAUX_RUN_ID__"`
+}
+
+type runtimeProperties struct {
+	RuntimeArgs *runtimeArgs `json:"runtimeArgs"`
+}
+
+func (ra *runtimeArgs) UnmarshalJSON(data []byte) error {
+	unquoted, err := strconv.Unquote(string(data))
+	if err != nil {
+		return fmt.Errorf("failed to escape runtime arguments %v: %v", string(data), err)
+	}
+
+	// Use alias to avoid infinite recursion: https://stackoverflow.com/q/52433467
+	type alias runtimeArgs
+	var a alias
+	if err := json.Unmarshal([]byte(unquoted), &a); err != nil {
+		return err
+	}
+	*ra = runtimeArgs(a)
+	return nil
+}
+
+// These are the only keys we need
+type run struct {
+	RunID      string `json:"runid"`
+	Status     string `json:"status"`
+	Properties runtimeProperties
```

Comment:
missing json tag? without the tag the json.Unmarshal call won't be able to set this field